### PR TITLE
chakrashim: implement Proxy methods

### LINKED ISSUE #488: util: util.types.isProxy does not work in Node-ChakraCore
Causes `test/parallel/test-util-types` to fail.

### LINKED ISSUE #523: util.inspect showProxy option doesn't work in node-chakra
The `util.inspect` option `showProxy` doesn't work in node-chakra

```js
util.inspect(new Proxy({},{}), { showProxy: true })
// "{}" but should be "Proxy [ {}, {} ]"
```
The [`showProxy` option](https://nodejs.org/dist/latest-v10.x/docs/api/util.html#util_util_inspect_object_options) was added in Node 6.x.

### PR DESCRIPTION
**Pull request by rhuanjl**
Implement: isProxy() Proxy::GetHandler() and Proxy::GetTarget()

As discussed in the issue log there are some test fails which to me appear to be unrelated to this change, any pointers appreciated otherwise I'll take another look tomorrow.
EDIT: the same tests fail with a clean checkout of master on macOS 10.13.4 I'll open a new issue for them.

First time contributing to node-chakracore (or node) so please forgive me if I done anything obviously stupid.

Fixes: https://github.com/nodejs/node-chakracore/issues/523
Fixes: https://github.com/nodejs/node-chakracore/issues/488
##### Checklist

- [ ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

### CODE REVIEW

File: deps/chakrashim/src/v8proxy.cc [hunk 1]
```diff
@@ -22,15 +22,23 @@
 
 namespace v8 {
 Local<Object> Proxy::GetTarget() {
-  // CHAKRA-TODO: Need to add JSRT API to get target of proxy
-  JsValueRef undefinedValue = jsrt::GetUndefined();
-  return Local<Object>::New(undefinedValue);
+  JsValueRef returnValue;
```
Comment (kfarnung) [inline deps/chakrashim/src/v8proxy.cc]: Can you just initialize this to `JS_INVALID_REFERENCE` to be safe?

File: deps/chakrashim/src/v8proxy.cc [hunk 2]
```diff
@@ -22,15 +22,23 @@
...
+  bool isProxy = false;
+  JsGetProxyProperties((JsValueRef) this, &isProxy, &returnValue, nullptr );
+  if (!isProxy || returnValue == nullptr) {
```
Comment (kfarnung) [inline deps/chakrashim/src/v8proxy.cc]: Prefer `JS_INVALID_REFERENCE` over `nullptr` even though they are the same for practical purposes.

**Comment by kfarnung** [comment]
Kicked off a CI:  https://ci.nodejs.org/job/chakracore-test-pull-request/246/

**Review by kfarnung — APPROVED** [review]
LGTM with nits

File: deps/chakrashim/src/v8proxy.cc [hunk 3]
```diff
@@ -22,15 +22,23 @@
 
 namespace v8 {
 Local<Object> Proxy::GetTarget() {
-  // CHAKRA-TODO: Need to add JSRT API to get target of proxy
-  JsValueRef undefinedValue = jsrt::GetUndefined();
-  return Local<Object>::New(undefinedValue);
+  JsValueRef returnValue;
+  bool isProxy = false;
+  JsGetProxyProperties((JsValueRef) this, &isProxy, &returnValue, nullptr );
```
Comment (jackhorton) [inline deps/chakrashim/src/v8proxy.cc]: nit: trailing space in argument list.

**Review by jackhorton — APPROVED** [review]

File: deps/chakrashim/src/v8proxy.cc [hunk 4]
```diff
@@ -22,15 +22,23 @@
...
+  return Local<Object>::New(returnValue);
 }
 
 Local<Value> Proxy::GetHandler() {
-  // CHAKRA-TODO: Need to add JSRT API to get handler of proxy
-  JsValueRef undefinedValue = jsrt::GetUndefined();
-  return Local<Value>::New(undefinedValue);
+  JsValueRef returnValue;
+  bool isProxy = false;
+  JsGetProxyProperties((JsValueRef) this, &isProxy, nullptr, &returnValue );
```
Comment (kfarnung) [inline deps/chakrashim/src/v8proxy.cc]: nit: trailing whitespace in arguments list here too

**Comment by rhuanjl** [comment]
Thanks for the reviews, I've pushed a commit that fixes the points raised, please let me know if I should squash it, also should this all be one commit or is the original 2 commit split a good idea?

**Comment by jackhorton** [comment]
We have to do squashing of our own on our end because of the super strict node commit guidelines (no merges ever allowed, specific form for commit messages).

**Comment by kfarnung** [comment]
I think it makes sense to just squash these ~~two~~three together.  Can you make sure the commit message conforms to the [node guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines) as well?

**Comment by rhuanjl** [comment]
Noticed something else all tests for this are currently marked as skip in parallel.status

Notably:
 - test.util.types (requires esmodule support - but that's only for one bit at the end of the file so could put an ```if(!common.isChakraEngine)``` around it and enable the test
 - test-util-inspect-proxy - test passes offline - simply tests what I've enabled here

To enable these do I just remove them from the parallel.status document?

(Sorry for not picking this up before I'm not used to the test interface)

**Comment by kfarnung** [comment]
Both of those approaches are correct.  If an entire test is skipped, just remove it from the `$jsEngine==chakracore` section of the status file.  If there's only part of a test which isn't supported, adding the `common.isChakraEngine` check is the right approach.  Adding a comment to the code indicating why we are skipping that part of the test is also a good idea.

Thanks for getting this all wired up and tested!

**Comment by rhuanjl** [comment]
Tests enabled, commits squashed and commit message updated.

**Review by kfarnung — APPROVED** [review]

**Comment by kfarnung** [comment]
Thanks! Kicked off a new CI: https://ci.nodejs.org/job/chakracore-test-pull-request/251/

**Comment by kfarnung** [comment]
Landed in 5d023bb320de120bab92f42f2bf3614b569e42bc

**Pull request closed without merging**